contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -0,0 +1,406 @@
+that are in "waiting" state. Once a Pod is approved, it is sent to the Bind
+stage.
+
+### Reject
+
+Plugins called at "Admit" may perform some operations that should be undone if
+the Pod reservation fails. The "Reject" extension point allows such clean-up
+operations to happen. Plugins registered at this point are called if the
+reservation of the Pod is cancelled. The reservation is cancelled if any of the
+"Admit" plugins returns "reject" or if a Pod reservation, which is in "wait"
+state, times out. 
+
+### Pre-Bind
+
+When a Pod is approved for binding it reaches to this stage. These plugins run
+before the actual binding of the Pod to a Node happens. The binding starts only
+if all of these plugins return true. If any returns false, the Pod is rejected
+and sent back to the scheduling queue. These plugins run in a separate go
+routine. The same go routine runs "Bind" after these plugins when all of them
+return true.
+
+### Bind
+
+Once all pre-bind plugins return true, the Bind plugins are executed. Multiple
+plugins may be registered at this extension point. Each plugin may return true
+or false (or an error). If a plugin returns false, the next plugin will be
+called until a plugin returns true. Once a true is returned **the remaining
+plugins are skipped**. If any of the plugins returns an error or all of them
+return false, the Pod is rejected and sent back to the scheduling queue.
+
+### Post Bind
+
+The Post Bind plugins can be useful for housekeeping after a pod is scheduled.
+These plugins do not return any value and are not expected to influence the
+scheduling decision made in the scheduling cycle.
+
+# USE-CASES
+
+In this section we provide a couple of examples on how the scheduling framework
+can be used to solve common scheduling scenarios.
+
+### Dynamic binding of cluster-level resources
+
+Cluster level resources are resources which are not immediately available on
+nodes at the time of scheduling Pods. Scheduler needs to ensure that such
+cluster level resources are bound to a chosen Node before it can schedule a Pod
+that requires such resources to the Node. We refer to this type of binding of
+resources to Nodes at the time of scheduling Pods as dynamic resource binding.  
+Dynamic resource binding has proven to be a challenge in Scheduler v1, because
+Scheduler v1 is not flexible enough to support various types of plugins at
+different phases of scheduling. As a result, binding of storage volumes is
+integrated in the scheduler code and some non-trivial changes are done to the
+scheduler extender to support dynamic binding of network GPUs.  
+The scheduling framework allows such dynamic bindings in a cleaner way. The main
+thread of scheduling framework process a pending Pod that requests a network
+resource and finds a node for the Pod and reserves the Pod. A dynamic resource
+binder plugin installed at "Pre-Bind" stage is invoked (in a separate thread).
```

Comment:
AFAIK, volume scheduling requires us to take volume details into consideration when making node chosen decision,
but seems we've already have a chosen node before "Pre-Bind" here. So, won't it be too late?
Or it turned out to be best-effort? i.e. try to bind to a chosen node, but not ensure it available before selecting a node.

Reply:
There are two scenarios:
1. a PV is already bound to a node and the Pod needs that PV: in this case a "filter" function must choose that node and filter out the rest of the nodes.
2. Pod requires a PV, but the PV is not bound to any node. Scheduler picks a node that satisfies other scheduling requirements of the Pod and then a pre-bind plugin is invoked to bind the PV to the chosen node. If the binding fails, the pre-bind plugin returns an error and the Pod goes back to the scheduling queue for another try. If binding succeeds, the Pod goes to the next stage and is bound to the node.

Reply:
My main concern is for the scenario 2: for now, we take volume binding/provisioning things into consideration during predicates, and scheduler will trigger binding/provisioning for unbound pvcs on a pod, to ensure the volume can work on the chosen node.

By "satisfies other scheduling requirements of the Pod", you mean we'll skip this in the V2 framework? Then there will be a chance that the Pre-bind plugin cannot find available volume to bind/provision for the chosen node. Or you'd suggest the scheduler keep retrying scheduling, until it finds available volume for the chosen node that time?

Also, currently scheduling will fail if there are unbound pvcs on a pod, I guess it will not step into the "Pre-Bind" stage.

Reply:
Or the plugin of predicate still exists, and the main difference lies in the "Reserve"?
Currently we may experience multi scheduling processes if the volume binding/provision is slow. And with the new mechanism, we can manage the thing in a same scheduling cycle?

Reply:
> By "satisfies other scheduling requirements of the Pod", you mean we'll skip this in the V2 framework?

No! I didn't mean that.

> there will be a chance that the Pre-bind plugin cannot find available volume to bind/provision for the chosen node. Or you'd suggest the scheduler keep retrying scheduling, until it finds available volume for the chosen node that time?

Yes, the scheduler will retry unschedulable pods and when there are volumes available, the pod may be scheduled.

> Also, currently scheduling will fail if there are unbound pvcs on a pod, I guess it will not step into the "Pre-Bind" stage.

Such pod could proceed to Pre-Bind and then binding can happen there, but as I mentioned above in my comment to Michelle, that logic has a potential race condition. So, we may want to consider a different design where reservation of a volume happens in earlier stages of a scheduling cycle.

Reply:
The latest design adds a "Reserve" phase that is similar to the current scheduler's "Assume" phase. I think this is sufficient for volume binding (along with predicates, priorities and bind phases)

Reply:
That's correct, the current predicate we have today maps to the filter phase of the scheduler v2